### ISSUE TITLE
["Request"] Specify `AT_MOST_ONCE` and `EXACTLY_ONCE` in applicable lambdas

### ISSUE THREAD

**Issue by nomisRev** [arrow-kt/arrow#2806]
In most HOF in Arrow we can specify `AT_MOST_ONCE` and `EXACTLY_ONCE`.
Currently, we're not doing so, and this prevents certain code to be written that would otherwise be valid.

**Comment by poseidon2060** [comment]
Hey, can I work on this issue?
I am relatively new to Android and I hope this can be a good beginner issue for me

**Comment by nomisRev** [comment]
Hey @poseidon2060,

Yes, this is a great issue to get started with since you can also start small and add `AT_MOST_ONCE` and `EXACTLY_ONCE` only in a couple or a single place at the same time. I would suggest getting familiar with Kotlin contracts and `callsInPlace`, and looking for any file in `libs/arrow-core/commonMain` and find any _lamda_ were you can add these annotations.

This doesn't have to be done in the entire project in 1 PR or contribution. Feel free to add it only to a single lambda where you think it's applicable and opening a PR 👍

**Comment by poseidon2060** [comment]
Okay thankyou
Can I ask here for my help and doubts, or do I have to join Slack?

**Comment by nomisRev** [comment]
Feel free to ask here @poseidon2060, or in a DRAFT PR if you prefer 👍

**Comment by poseidon2060** [comment]
Hey @nomisRev 

I have familiarized myself with Kotlin contracts and `callsInPlace` but I am unable to figure out where to apply them in the project
I browsed the path that you mentioned but am unable to understand exactly where I can add these annotations
Can you guide me slightly on what to look for, help appreciated

**Comment by nomisRev** [comment]
Hey @poseidon2060,

If we look for example at `Either`, there are many places where `AT_MOST_ONCE` can be added. I.e. for [`fold`](https://github.com/arrow-kt/arrow/blob/a43965c001b334c0a0ec794dc31fd1a9ee013e9c/arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Either.kt#L826) you know that _either_ the function for `ifLeft` or `ifRight` will be called `AT_MOST_ONCE`

To add this annotations for `Either.kt`, you need to inspect the method body and see if the `lambda` is called `AT_LEAST_ONCE` or `AT_MOST_ONCE` and add that into the method body. This should only be done for methods that are not `@Deprecated`.

If you any more questions be sure to ask them ☺️ I hope this helps.

**Comment by poseidon2060** [comment]
Hey, @nomisRev 

Thank you for being so helpful!
I tried out what you said for [`fold`](https://github.com/poseidon2060/arrow/blob/a0414b158dbe8d9e318e93377058dcdf25897514/arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Either.kt#L826) and this is what I changed

```
@OptIn(ExperimentalContracts::class)
  public inline fun <C> fold(ifLeft: (left: A) -> C, ifRight: (right: B) -> C): C =
    when (this) {
      is Right -> {
        contract { callsInPlace(ifRight,InvocationKind.AT_MOST_ONCE) }
        ifRight(value)
      }
      is Left -> {
        contract { callsInPlace(ifLeft,InvocationKind.AT_MOST_ONCE) }
        ifLeft(value)
      }
    }
```
Am I doing it correctly?
I hope I didn't do something terribly wrong😅

**Comment by hoc081098** [comment]
I think should move all contract blocks to the top body of function.

```
contract {
  callInPlace(ifRight, 
...)
  callInPlace(ifLeft, ...)
}
when(this) { ... }
```

**Comment by poseidon2060** [comment]
> I think should move all contract blocks to the top body of function.
> 
> ```
> contract {
>   callInPlace(ifRight, 
> ...)
>   callInPlace(ifLeft, ...)
> }
> when(this) { ... }
> ```

But I don't think we can add contracts outside the function body. Or am I missing something?

**Comment by poseidon2060** [comment]
> 

Do you mean like this?
```
@OptIn(ExperimentalContracts::class)
  public inline fun <C> fold(ifLeft: (left: A) -> C, ifRight: (right: B) -> C): C {
    contract {
      callsInPlace(ifLeft, InvocationKind.AT_MOST_ONCE)
      callsInPlace(ifRight, InvocationKind.AT_MOST_ONCE)
    }
    return when (this) {
      is Right -> ifRight(value)
      is Left -> ifLeft(value)
    }
  }
```

**Comment by poseidon2060** [comment]
Hey @hoc081098 

Do you mean like this?

```
@OptIn(ExperimentalContracts::class)
  public inline fun <C> fold(ifLeft: (left: A) -> C, ifRight: (right: B) -> C): C {
    contract {
      callsInPlace(ifLeft, InvocationKind.AT_MOST_ONCE)
      callsInPlace(ifRight, InvocationKind.AT_MOST_ONCE)
    }
    return when (this) {
      is Right -> ifRight(value)
      is Left -> ifLeft(value)
    }
  }
```

**Comment by nomisRev** [comment]
Yes, the second snippet as suggested by @hoc081098 is the correct encoding. `contract` needs to be the first statement in the function body.

**Comment by poseidon2060** [comment]
> Yes, the second snippet as suggested by @hoc081098 is the correct encoding. `contract` needs to be the first statement in the function body.

So can I open an PR?

**Comment by nomisRev** [comment]
Absolutely ☺️

**Comment by poseidon2060** [comment]
Thank you @nomisRev and @hoc081098 for being so helpful😊

**Comment by poseidon2060** [comment]
Hey @nomisRev, I have created a Draft PR adding Invocation at 4 more places, can you please let me know if it's correct so that I can mark it Ready for Review🤞

**Comment by poseidon2060** [comment]
Hey @nomisRev, should I continue working on this issue or get started on a new one?
I don't know if I am supposed to complete this for all the files or not, I'd do it if you want🤔

**Comment by nomisRev** [comment]
Hey @poseidon2060,
This needs to happen for all files (you can ignore `@Deprecated` code but also feel free to work on an other issue if you like ☺️

**Comment by nomisRev** [comment]
@poseidon2060 awesome 🙌 Thank you for taking so much interest to contributing!
These might be some interesting tickets to look into:
 - https://github.com/arrow-kt/arrow/issues/2908
 - https://github.com/arrow-kt/arrow/issues/2894

Feel free to ask any questions on the respective issues directly, or on any other issues you find interesting and tag me 😉
I'll do my best to add some more beginner friendly tickets in the coming weeks.

**Comment by poseidon2060** [comment]
@nomisRev Okay thank you, I will look into these issues and try to solve them
Contributing to open source is really fun, I am glad I started with this repository😊

**Comment by serras** [comment]
Closed by and #2901 and #2915
